Update omero-grid to the latest server

manics opened the pull request:
See the [README.md](https://github.com/manics/ome-docker/blob/omero-grid-new/omero-grid/README.md) for instructions

--
Ignore the rest of this description, it should be fixed by https://github.com/openmicroscopy/openmicroscopy/pull/4422

Unfortunately this doesn't work, see https://trello.com/c/3fY2taKP/88-omeor-grid-omero-node-bugs-issues

Attempting to start the slave (Processor-0) leads to an initial Ice connection, but `Processor-0.log` contains
```
2015-11-22 19:03:33,716 INFO  [                       omero.util.Server] (MainThread) ********************************************************************************
2015-11-22 19:03:33,716 INFO  [                       omero.util.Server] (MainThread) Waiting 10000 ms on startup
2015-11-22 19:03:43,717 INFO  [                       omero.util.Server] (MainThread) Starting
2015-11-22 19:03:43,719 INFO  [                    omero.util.Resources] (Thread-2  ) Starting
2015-11-22 19:03:43,791 ERROR [                       omero.util.Server] (MainThread) Failed activation
Traceback (most recent call last):
  File "/home/omero/OMERO.server-5.2.0-ice35-b12/lib/python/omero/util/__init__.py", line 502, in run
    self.adapter_name)
  File "/usr/lib64/python2.7/site-packages/Ice/Ice.py", line 557, in createObjectAdapter
    adapter = self._impl.createObjectAdapter(name)
SocketException: Ice.SocketException:
Cannot assign requested address
2015-11-22 19:03:43,794 INFO  [                       omero.util.Server] (MainThread) Cleanup
2015-11-22 19:03:43,794 INFO  [              omero.processor.ProcessorI] (MainThread) Cleaning up
2015-11-22 19:03:43,794 INFO  [              omero.processor.ProcessorI] (MainThread) Done
2015-11-22 19:03:43,821 WARNI [                omero.util.ServerContext] (MainThread) Connection failure: exception ::Ice::ObjectNotExistException
{
    id = 
    {
        name = 3505efce-99d6-4f5f-b827-42e87f4278da
        category = session-eb45540f-1e4c-4d4c-b02a-447908ebc69b
    }
    facet = 
    operation = keepAlive
}
2015-11-22 19:03:43,821 INFO  [                       omero.util.Server] (MainThread) Stopped
2015-11-22 19:03:43,821 INFO  [                       omero.util.Server] (MainThread) ********************************************************************************
2015-11-22 19:03:43,822 WARNI [                                  stderr] (MainThread) 2015-11-22 19:03:43.821998 Processor-0: error: Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/Ice/Ice.py", line 1127, in main
    status = self.doMain(args, initData)
  File "/usr/lib64/python2.7/site-packages/Ice/Ice.py", line 1174, in doMain
    return self.run(args)
  File "/home/omero/OMERO.server-5.2.0-ice35-b12/lib/python/omero/util/__init__.py", line 512, in run
    sys.exit(200)
SystemExit: 200
2015-11-22 19:03:43,841 INFO  [                    omero.util.Resources] (Thread-2  ) Halted
```

TODO:
- [x] Remove old `db` image
- [x] Remove old `omero-server` and `omero-slave` images (`omero` image does both)
- [x] Update `omero-web`
- [x] Cleanup

manics commented:
I think this is done, apart from some cleanups e.g. removing web dependencies from `omero/Dockerfile` (I could do this here or in a later PR). See the README: https://github.com/manics/ome-docker/tree/omero-grid-new/omero-grid

manics commented:
This might be an interesting test:
```
# Glacier only
docker run -d --link postgres:db \
    -e DBUSER=postgres -e DBPASS=postgres -e DBNAME=postgres \
    -p 4063:4063 -p 4064:4064 \
    --name omero-master omero-grid master \
    slave-b0:Blitz-0 slave-b1:Blitz-1 \
    slave-i0:Indexer-0 slave-i1:Indexer-1 \
    slave-d:DropBox,MonitorServer,FileServer slave-s:Storm \
    slave-pi0:PixelData-0 slave-pi1:PixelData-1 \
    slave-t0:Tables-0 slave-t1:Tables-1 \
    slave-pr0:Processor-0 slave-pr1:Processor-1

for slave in slave-b0 slave-b1 slave-i0 slave-i1 slave-d slave-pi0 slave-pi1 \
        slave-t0 slave-t1 slave-pr0 slave-pr1; do
    docker run -d --link omero-master:master \
        --volumes-from omero-master \
        --name $slave omero-grid $slave
done

docker run -d --link omero-master:master -p 80:8080 omero-grid-web
```

joshmoore commented on omero-grid/omero/Dockerfile:
```diff
@@ -1,34 +1,49 @@
-    ln -s OMERO.server-5.0.1-ice35-b21 OMERO.server"
+USER omero
+WORKDIR /home/omero
+RUN omego download server --release $OMERO_VERSION && \
```
This will likely be the first thing that I'd need to change. Can we make all the arguments passable? Or unpack to a well-known location so an unzipped server can be mounted?

manics replied:
How would you feel about a build-arg for the ci-server and branch?

joshmoore replied:
Would work for me. Could be part of the tag that we use: e.g. `docker build -t omero-deploy:b148 --build-arg ...`

manics replied:
That's exactly what I was thinking... have a script which takes the version args, sets build-args and somehow munges them into a tag

joshmoore commented:
Cherry-pick https://github.com/joshmoore/ome-docker/commit/d77120ad35f8cd50da2dac0961c4ee383723541a ?

manics commented:
Cherry-picked, `./build.py` added to pass the CI server and branch to the build, README updated.
- Discovered a mysterious bug in `omego download --ci ...`, the `ci` argument appears to be ignored (but it is respected in `omego upgrade`.
- Travis Docker is too old and doesn't support `--build-arg`

joshmoore commented:
From https://blog.travis-ci.com/2015-08-19-using-docker-on-travis-ci/ :

> The default Docker version installed is currently 1.7.0. We'll keep it up-to-date as we roll out new build environments, but you're also free to install different versions of Docker as part your builds, the APT repository is added by default.

manics commented:
@joshmoore If you want to play with 5.1 this week you'll need to comment out the rewrite workarounds in https://github.com/ome/ome-docker/pull/24/files#diff-7fe777f48c77321016945317c474b5df

manics commented:
The current state of this is:

- All in one server + web works with the metadata branch:
  - `./build.py --ci metadata-ci.docker.openmicroscopy.org:8080 --release OMERO-build omero-grid...`
  - `docker run ... metadata-ci-docker-openmicroscopy-org-8080/omero-grid:omero-build`
  - `docker run ... metadata-ci-docker-openmicroscopy-org-8080/omero-grid-web:omero-build`

- Multiple server nodes works with OMERO-DEV-latest:
  - `./build.py --release OMERO-DEV-latest omero-grid...`
  - `docker run -d --link xxx-db:db -e DBUSER=postgres -e DBPASS=postgres -e DBNAME=postgres --name xxx-master openmicroscopy/omero-grid:omero-dev-latest master master:Blitz-0,Indexer-0,DropBox,MonitorServer,FileServer,Storm,PixelData-0,Tables-0 slave-1:Processor-0`
  - `docker run -d --link xxx-master:master --name xxx-slave1 openmicroscopy/omero-grid:omero-dev-latest slave-1`
  - `docker run ... openmicroscopy/omero-grid:omero-dev-latest master`

manics commented:
@joshmoore Preferences on volumes? Currently this is setup so `/OMERO/`, `/home/omero/OMERO.server/var` and `/home/omero/nginx` are defined as volumes, I could keep this, move everything to a single `/data` as before, or keep them separate but move to some other location.

joshmoore commented:
is `nginx` running as `omero`? If not, I'd think one mount for `nginx` and then either the other dirs as you have them or a single `/data` for the `omero` user would be fine.

joshmoore commented:
Is there any benefit of getting these images onto hub (considering the complexity of `build.py`)?

manics commented:
That's my plan, trying to do too many other things at the same time.

manics commented:
@joshmoore I think this is ready for testing.

joshmoore commented:
Blocker (or PR doc issue?):

 * `./build.py omero-grid`
 * `docker tag openmicroscopy/omero-grid:latest omero-grid:josh-test-24`
 * `docker run omero-grid:josh-test-24` exits

```
[jamoore@orca-3 omero-grid]$ docker run omero-grid:josh-test-24
Entering a shell
[jamoore@orca-3 omero-grid]$ docker ps
```

The readme says to add "master" which works.

----

Few non-blocker comments (i.e. they don't prevent us getting to production with this but affect usability):
 * `build.py` doesn't provide a way to name the image meaning that it's fairly easy to overwrite someone else's run
 * Use of build-args for CI look prevents use of links to download from a container on the same host (ideally I'd like to have the build in a docker-compose)


----

Testing continues after lunch.

manics commented:
> Blocker (or PR doc issue?):

Originally the default was to drop into a bash shell which in hindsight is confusing if you're not running interactively. Changed to `master` now , so no args or just `master` should start a full server (minus web).

>  build.py doesn't provide a way to name the image

Added `./build --tag` to override the full name/tag

> Use of build-args for CI look prevents use of links to download from a container on the same host 

Do you mean you're unable to link directly to a zip? If so then that's because I'm using `omego download` instead of `omego install`, otherwise OMERO.server automatically starts as a daemon. See https://github.com/ome/omego/pull/69

joshmoore commented:
> Do you mean you're unable to link directly to a zip? If so then that's because I'm using omego download instead of omego install, otherwise OMERO.server automatically starts as a daemon. See ome/omego#69

No, I mean due to DNS, I don't have an IP (in general) for my CI, but I *can* easily link it into a container for us. But linking isn't available at build time.

manics commented:
Last three commits should help with testing. Runs a master-slave config, and you can modify params at the top of `test.sh`, e.g. see https://github.com/openmicroscopy/openmicroscopy/pull/4422

manics commented:
> No, I mean due to DNS, I don't have an IP (in general) for my CI, but I can easily link it into a container for us. But linking isn't available at build time.

So that's going back to the omero-deploy workflow of installing at runtime? How about a separate PR? I'll could add a `--build-arg devel` flag to specify the server is setup during `run` not `build`

joshmoore commented:
Summary:

 * works based on https://github.com/openmicroscopy/openmicroscopy/pull/4422#issuecomment-176045095 (but see the minor RFE there)
 * for the moment, this is certainly working, so yes, let's handle `build-arg` later. For the moment, I can likely hard-code the IP somewhere to get my builds to work.
 * there might be room for simplification in the little language you're using for service configuration but that's basically coming up with a new API, so definitely down the road.

joshmoore commented:
https://travis-ci.org/ome/ome-docker/builds/105129380 is also red.

manics commented:
https://s3.amazonaws.com/archive.travis-ci.org/jobs/105445281/log.txt

- build.py checks env-vars (I've defaulted to empty so that it matches the default behaviour of build.py and overriden the env-var in `.travis.yml`, though if you disagree I'll hard-code it in `test.sh`
- Added smoke tests for web login, dropbox and processor

manics commented:
I think I really am done now.

joshmoore commented:
I can successfully get a 5.2 server up and running now. :+1: (That *doesn't* help me with the metadata51 build, though, but perhaps that's just motivation for getting us migrated)

manics commented:
This was stretching out so I decided to focus on 5.2 develop. Now that it's in I could work on `metadata51` support if the rebase is going to take a while, add a trello card if you want me to (and find a column for it...).

joshmoore commented:
My primary need is to test @mtbc's DB, but I may be able to do that with an upgraded version. I'll try that first since we'll need to upgrade anyway.
